<?xml version="1.0"?>
<robot name="kuka-iiwa-7-egp-40" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- Fix to world just for testing -->
  <link name="world"/>
  <!--joint between {parent} and iiwa7_link_0-->
  <joint name="world_iiwa_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="world"/>
    <child link="iiwa7_link_0"/>
  </joint>

  <link name="iiwa7_link_0">
    <inertial>
      <origin xyz="-0.01 0 0.07" rpy="0 0 0"/>
      <mass value="5"/>
      <inertia ixx="0.05"  ixy="0"  ixz="0" iyy="0.06" iyz="0" izz="0.03" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_0.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_0.stl"/>
      </geometry>
    </collision>

    <self_collision_checking>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <capsule radius="0.15" length="0.25"/>
      </geometry>
    </self_collision_checking>

  </link>

  <!-- joint between iiwa7_link_0 and iiwa7_link_1 -->
  <joint name="iiwa7_joint_1" type="revolute">
    <parent link="iiwa7_link_0"/>
    <child link="iiwa7_link_1"/>
    <origin xyz="0 0 0.157" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.96705972839" upper="2.96705972839"
           effort="300" velocity="1.71042266695" />
    <safety_controller soft_lower_limit="-2.93215314335"
                       soft_upper_limit="2.93215314335"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_1">
    <inertial>
      <origin xyz="0 -0.03 0.12" rpy="0 0 0"/>
      <mass value="3.4525"/>
      <inertia ixx="0.02183"  ixy="0"  ixz="0" iyy="0.02081" iyz="0.003883" izz="0.007697" />
    </inertial>
    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_1.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_1.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_2" type="revolute">
    <parent link="iiwa7_link_1"/>
    <child link="iiwa7_link_2"/>
    <origin xyz="0 0 0.183" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-2.09439510239" upper="2.09439510239"
           effort="300" velocity="1.71042266695" />
    <safety_controller soft_lower_limit="-2.05948851735"
                       soft_upper_limit="2.05948851735"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_2">
    <inertial>
      <origin xyz="0.000 0.034417 0.0673286" rpy="0 0 0"/>
      <mass value="3.4821"/>
      <inertia ixx="0.02177"  ixy="0"  ixz="0.00" iyy="0.02075" iyz="0.003621" izz="0.00779" />
    </inertial>
    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_2.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_2.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_3" type="revolute">
    <parent link="iiwa7_link_2"/>
    <child link="iiwa7_link_3"/>
    <origin xyz="0 0 0.185" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.96705972839" upper="2.96705972839"
           effort="300" velocity="1.74532925199" />
    <safety_controller soft_lower_limit="-2.93215314335"
                       soft_upper_limit="2.93215314335"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_3">
    <inertial>
      <origin xyz="0 0.0295 0.1258" rpy="0 0 0"/>
      <mass value="4.05623"/>
      <inertia ixx="0.03204"  ixy="0"  ixz="0" iyy="0.03044" iyz="-0.006231" izz="0.009726" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_3.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_3.stl"/>
      </geometry>
    </collision>

  </link>

  <joint name="iiwa7_joint_4" type="revolute">
    <parent link="iiwa7_link_3"/>
    <child link="iiwa7_link_4"/>
    <origin xyz="0 0 0.215" rpy="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit lower="-2.09439510239" upper="2.09439510239"
           effort="300" velocity="2.26892802759" />
    <safety_controller soft_lower_limit="-2.05948851735"
                       soft_upper_limit="2.05948851735"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_4">
    <inertial>
      <origin xyz="0 -0.034 0.067" rpy="0 0 0"/>
      <mass value="3.4822"/>
      <inertia ixx="0.021775"  ixy="0"  ixz="0" iyy="0.02075" iyz="-0.003621" izz="0.007783" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_4_y_z.obj"/>
      </geometry>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_4.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_5" type="revolute">
    <parent link="iiwa7_link_4"/>
    <child link="iiwa7_link_5"/>
    <origin xyz="0 0 0.4" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.96705972839" upper="2.96705972839"
           effort="300" velocity="2.44346095279" />
    <safety_controller soft_lower_limit="-2.93215314335"
                       soft_upper_limit="2.93215314335"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_5">
    <inertial>
      <origin xyz="0.0001 -0.02263 -0.1365" rpy="0 0 0"/>
      <mass value="2.1633"/>
      <inertia ixx="0.01287"  ixy="0"  ixz="0" iyy="0.01112" iyz="0.003943" izz="0.0057128" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_5.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_5.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_6" type="revolute">
    <parent link="iiwa7_link_5"/>
    <child link="iiwa7_link_6"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-2.09439510239" upper="2.09439510239"
           effort="300" velocity="3.14159265359" />
    <safety_controller soft_lower_limit="-2.05948851735"
                       soft_upper_limit="2.05948851735"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_6">
    <inertial>
      <origin xyz="0 0.00047 0.002124" rpy="0 0 0"/>
      <mass value="2.3466"/>
      <inertia ixx="0.006522"  ixy="0"  ixz="0" iyy="0.006270" iyz="0.00032280" izz="0.0045345" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_6.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_6.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_7" type="revolute">
    <parent link="iiwa7_link_6"/>
    <child link="iiwa7_link_7"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.05432619099" upper="3.05432619099"
           effort="300" velocity="3.14159265359" />
    <safety_controller soft_lower_limit="-3.01941960595"
                       soft_upper_limit="3.01941960595"
                       k_position="100"
                       k_velocity="2"/>
    <dynamics damping="0.5"/>
  </joint>

  <link name="iiwa7_link_7">
    <inertial>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <mass value="3.129"/>
      <inertia ixx="0.0022711"  ixy="0.0000221"  ixz="0" iyy="0.0022807" iyz="0" izz="0.002872" />
    </inertial>

    <visual>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/visual/link_7.dae"/>
      </geometry>
    </visual>

    <collision>
      <geometry>
        <mesh filename="model://kuka-iiwa-7/meshes/collision/link_7.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="iiwa7_joint_ee" type="fixed">
    <parent link="iiwa7_link_7"/>
    <child link="SchunkEGP40_link"/>
    <origin xyz="0 0 0.126" rpy="0 0 0"/>
  </joint>

  <!-- SCHUNK EGP 40 EEF -->

  <!-- Base -->
    <link name="SchunkEGP40_link">
        <inertial>
            <mass value="0.6" />
            <origin rpy="0 0 0" xyz="0.00078059 -0.00070996 0.04726637" />
            <inertia ixx="0.0011357" ixy="-0.00000024461" ixz="0.000014912" iyy="0.0012832" iyz="0.0000085651" izz="0.00066545" />
        </inertial>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0" />
            <geometry>
                <mesh filename="model://schunk-egp-40/meshes/visual/obj/schunk_low_baked_small.obj" scale="1 1 1" />
            </geometry>
            <material name="white">
                <color rgba="1 1 1 1"/>
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0" />
            <geometry>
                <mesh filename="model://schunk-egp-40/meshes/visual/dae/schunk-egp-40-collision.dae" scale="1 1 1" />
            </geometry>
        </collision>
    </link>

    <!-- Finger 1 -->
    <link name="SchunkEGP40_Finger1_link">
        <inertial>
            <mass value="0.04" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.0000013454" ixy="0" ixz="0" iyy="0.00000061207" iyz="0" izz="0.00000097333" />
        </inertial>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0" />
            <geometry>
                <mesh filename="model://schunk-egp-40/meshes/visual/dae/schunk-egp-40-finger-base.dae" scale="1 1 1" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0.0005 0.006075" />
            <geometry>
                <box size="0.006 0.016 0.01215" />
            </geometry>
        </collision>
    </link>

    <!-- Finger 2 -->
    <link name="SchunkEGP40_Finger2_link">
        <inertial>
            <mass value="0.04" />
            <origin rpy="0 0 0" xyz="0 0 0" />
            <inertia ixx="0.0000013454" ixy="0" ixz="0" iyy="0.00000061207" iyz="0" izz="0.00000097333" />
        </inertial>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0" />
            <geometry>
                <mesh filename="model://schunk-egp-40/meshes/visual/dae/schunk-egp-40-finger-base.dae" scale="1 1 1" />
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0.0005 0.006075" />
            <geometry>
                <box size="0.006 0.016 0.01215" />
            </geometry>
        </collision>
    </link>

    <!-- Finger Joint 1 -->
    <joint name="SchunkEGP40_Finger1_joint" type="prismatic">
        <origin xyz="0.006800 0.000500 0.108000" rpy="0 0 0" />
        <axis xyz="1 0 0" />
        <limit effort="1" lower="0" upper="0.01" velocity="-1" />
        <dynamics damping="1.0" />
        <parent link="SchunkEGP40_link" />
        <child link="SchunkEGP40_Finger1_link" />
    </joint>
    <gazebo reference="SchunkEGP40_Finger1_joint">
        <provideFeedback>1</provideFeedback>
        <implicitSpringDamper>1</implicitSpringDamper>
    </gazebo>
    
    <!-- Finger Joint 2 -->
    <joint name="SchunkEGP40_Finger2_joint" type="prismatic">
        <origin xyz="-0.006800 0.000500 0.108000" rpy="0 0 3.14" />
        <axis xyz="1 0 0" />
        <limit effort="1" lower="0" upper="0.01" velocity="-1" />
        <dynamics damping="1.0" />
        <parent link="SchunkEGP40_link" />
        <child link="SchunkEGP40_Finger2_link" />
    </joint>
    <gazebo reference="SchunkEGP40_Finger2_joint">
        <provideFeedback>1</provideFeedback>
        <implicitSpringDamper>1</implicitSpringDamper>
    </gazebo>

  <!--Extensions -->
  <!-- Link0 -->
  <gazebo reference="iiwa7_link_0">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link1 -->
  <gazebo reference="iiwa7_link_1">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link2 -->
  <gazebo reference="iiwa7_link_2">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link3 -->
  <gazebo reference="iiwa7_link_3">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link4 -->
  <gazebo reference="iiwa7_link_4">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link5 -->
  <gazebo reference="iiwa7_link_5">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link6 -->
  <gazebo reference="iiwa7_link_6">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <!-- Link7 -->
  <gazebo reference="iiwa7_link_7">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
</robot>
